Add more ECC dispatch to a driver testing

ronald-cron-arm opened the pull request:
## Description
This PR progresses https://github.com/Mbed-TLS/TF-PSA-Crypto/issues/62.
Depends on: ~~#612~~

## PR checklist
- [x] **changelog** not required because: additional testing 
- [x] **framework PR** not required
- [x] **mbedtls development PR** not required because: TF-PSA-Crypto only work 
- [x] **mbedtls 3.6 PR** not required because: TF-PSA-Crypto only work
- **tests**  provided: additional crypto tests

ronald-cron-arm committed:
Add EC-JPAKE acceleration component

Equivalent of test_psa_crypto_config_accel_pake
in Mbed TLS.

Signed-off-by: Ronald Cron <ronald.cron@arm.com>

ronald-cron-arm committed:
Add ECC algs plus some key types acceleration component

Equivalent of test_psa_crypto_config_accel_ecc_some_key_types
in Mbed TLS.

Signed-off-by: Ronald Cron <ronald.cron@arm.com>

ronald-cron-arm committed:
Add some elliptic curves acceleration components

Equivalent of test_psa_crypto_config_accel_ecc_weierstrass_curves
and test_psa_crypto_config_accel_ecc_non_weierstrass_curves
in Mbed TLS

Signed-off-by: Ronald Cron <ronald.cron@arm.com>

ronald-cron-arm committed:
Add all ECC acceleration component

Equivalent of test_psa_crypto_config_accel_ecc_no_ecp_at_all
in Mbed TLS.

Signed-off-by: Ronald Cron <ronald.cron@arm.com>

ronald-cron-arm committed:
Add all ECC but ECP_LIGHT acceleration component

Equivalent of test_psa_crypto_config_accel_ecc_ecp_light_only
in Mbed TLS.

Signed-off-by: Ronald Cron <ronald.cron@arm.com>

ronald-cron-arm committed:
Add no bignum components

Equivalent of test_psa_crypto_config_accel_no_bignum
and test_psa_crypto_config_accel_ecc_ffdh_no_bignum
in Mbed TLS.

Signed-off-by: Ronald Cron <ronald.cron@arm.com>

ronald-cron-arm committed:
Move mbedtls_psa_key_agreement_iop_xyz declaration in psa_crypto_cipher.h

Signed-off-by: Ronald Cron <ronald.cron@arm.com>

ronald-cron-arm committed:
ecp: Switch to PSA error codes except for MBEDTLS_ERR_ECP_INVALID_KEY

Instead of using legacy error codes and translating
them with `mbedtls_to_psa_error()` to PSA
error codes, use the PSA error codes directly.
Do not replace the legacy macros, only change
their value. This is intended to avoid complicating
backports to 3.6.

Do not aplly this change for MBEDTLS_ERR_ECP_INVALID_KEY
for the time being as there is a tricky case in
`psa_crypto_pake.c` that requires non-trivial investigation.

The new value for MBEDTLS_ERR_ECP_INVALID_KEY would be
that of PSA_ERROR_INVALID_ARGUMENT but in `psa_crypto_pake.c`
it is also translated to PSA_ERROR_DATA_INVALID.

Moreover, we cannot simply translate PSA_ERROR_INVALID_ARGUMENT
to PSA_ERROR_DATA_INVALID in `psa_crypto_pake.c`, because
MBEDTLS_ERR_ECP_BAD_INPUT_DATA also maps to
PSA_ERROR_INVALID_ARGUMENT, and we do not want to translate
in that case to PSA_ERROR_DATA_INVALID.

Signed-off-by: Ronald Cron <ronald.cron@arm.com>

ronald-cron-arm committed:
Rework ECDH and ECDSA acceleration components

Use user-config-accel-ecc.h instead of a specific
one to reduce the number of ECC related user config
files.

Signed-off-by: Ronald Cron <ronald.cron@arm.com>

ronald-cron-arm committed:
psa_crypto.c: Remove guard for MBEDTLS_ERR_ECP_INVALID_KEY translation

Remove guard for MBEDTLS_ERR_ECP_INVALID_KEY translation
so that it is also available for the generated test driver.

Signed-off-by: Ronald Cron <ronald.cron@arm.com>

ronald-cron-arm commented:
Only the 10 last commits belong to this PRs. The eleven first belong to https://github.com/Mbed-TLS/TF-PSA-Crypto/pull/612

ronald-cron-arm committed:
Move mbedtls_ct_mpi_uint_if* static inline functions

Move mbedtls_ct_mpi_uint_if* static inline
functions that are bignum specific to
bignum_internal.h.

Signed-off-by: Ronald Cron <ronald.cron@arm.com>

minosgalanakis commented on tests/scripts/components-configuration-crypto.sh:
```diff
@@ -11,20 +11,347 @@
+    # Disable all the features that auto-enable ECP_LIGHT (see build_info.h)
+    scripts/config.py unset MBEDTLS_PK_PARSE_EC_EXTENDED
+    scripts/config.py unset MBEDTLS_PK_PARSE_EC_COMPRESSED
+    scripts/config.py unset PSA_WANT_KEY_TYPE_ECC_KEY_PAIR_DERIVE
```
It looks like this is new. In the mbedtls component we are not setting `MBEDTLS_PSA_ACCEL_KEY_TYPE_ECC_KEY_PAIR_DERIVE`

Similarly `MBEDTLS_PSA_ACCEL_ECC_SECP_K1_192`  `MBEDTLS_PSA_ACCEL_ECC_SECP_R1_192` are being accelerated in the mbedtls side but not here

ronald-cron-arm replied:
This is the equivalent of the Mbed TLS component `test_psa_crypto_config_accel_ecc_no_ecp_at_all` where MBEDTLS_PSA_ACCEL_KEY_TYPE_ECC_KEY_PAIR_DERIVE is unset (see config_psa_crypto_no_ecp_at_all).
Otherwise the support for `MBEDTLS_PSA_ACCEL_ECC_SECP_K1_192` and ` MBEDTLS_PSA_ACCEL_ECC_SECP_R1_192` has been removed in TF-PSA-Crypto.

minosgalanakis reviewed (commented):
Looks good but will need to rebase and flix the conflicts now. 

I have manually confirmed the configuration used in the component migration and the inclusion of the `tests/configs/user-config-accel-ecc.h` <-- 'user-config-test-driver-extension.h' against the old MbedTLS macros e.g `config_psa_crypto_config_ecp_light_only`. 

I can confirm there is feature parity, with only two exceptions which I commented bellow and are as intended. 

Will approve if we get a green CI.

davidhorstmann-arm commented on drivers/builtin/src/psa_crypto_pake.c:
```diff
@@ -112,10 +112,6 @@ static psa_status_t mbedtls_ecjpake_to_psa_error(int ret)
     switch (ret) {
         case MBEDTLS_ERR_ECP_INVALID_KEY:
             return PSA_ERROR_DATA_INVALID;
-        case MBEDTLS_ERR_MD_FEATURE_UNAVAILABLE:
-            return PSA_ERROR_NOT_SUPPORTED;
```
Why is this one removed? This error value is still the legacy error value:
```c
#define MBEDTLS_ERR_MD_FEATURE_UNAVAILABLE                -0x5080
```

ronald-cron-arm replied:
Good catch, leftover from some experimentation it seems.

davidhorstmann-arm commented on tests/configs/user-config-accel-ecc-ffdh.h:
```diff
@@ -0,0 +1,34 @@
+#define MBEDTLS_PSA_ACCEL_ECC_BRAINPOOL_P_R1_512
+#define MBEDTLS_PSA_ACCEL_ECC_MONTGOMERY_255
+#define MBEDTLS_PSA_ACCEL_ECC_MONTGOMERY_448
+#define MBEDTLS_PSA_ACCEL_ECC_SECP_K1_192
```
As before, we've removed this option.

davidhorstmann-arm commented on tests/configs/user-config-accel-ecc-ffdh.h:
```diff
@@ -0,0 +1,34 @@
+#define MBEDTLS_PSA_ACCEL_ECC_MONTGOMERY_448
+#define MBEDTLS_PSA_ACCEL_ECC_SECP_K1_192
+#define MBEDTLS_PSA_ACCEL_ECC_SECP_K1_256
+#define MBEDTLS_PSA_ACCEL_ECC_SECP_R1_192
```
Ditto

davidhorstmann-arm commented on tests/configs/user-config-accel-ecc.h:
```diff
@@ -0,0 +1,22 @@
+#define MBEDTLS_PSA_ACCEL_ECC_BRAINPOOL_P_R1_512
+#define MBEDTLS_PSA_ACCEL_ECC_MONTGOMERY_255
+#define MBEDTLS_PSA_ACCEL_ECC_MONTGOMERY_448
+#define MBEDTLS_PSA_ACCEL_ECC_SECP_K1_192
```
This was removed in #568 / #570.

davidhorstmann-arm commented on tests/configs/user-config-accel-ecc.h:
```diff
@@ -0,0 +1,22 @@
+#define MBEDTLS_PSA_ACCEL_ECC_MONTGOMERY_448
+#define MBEDTLS_PSA_ACCEL_ECC_SECP_K1_192
+#define MBEDTLS_PSA_ACCEL_ECC_SECP_K1_256
+#define MBEDTLS_PSA_ACCEL_ECC_SECP_R1_192
```
Ditto

davidhorstmann-arm commented on tests/scripts/components-configuration-crypto.sh:
```diff
@@ -11,6 +11,45 @@
+    cd $OUT_OF_SOURCE_DIR
+    cmake -DTF_PSA_CRYPTO_TEST_DRIVER=On \
+          -DTF_PSA_CRYPTO_USER_CONFIG_FILE="../tests/configs/user-config-accel-ecc.h" ..
+    make
```
Nit: Could do `cmake --build .` for compatibility with other generators.

davidhorstmann-arm commented on tests/scripts/components-configuration-crypto.sh:
```diff
@@ -91,6 +91,47 @@ component_test_accel_ecc_all_but_ecp_light() {
+    # the future, the following line could be removed (see issues
+    # 6061, 6332 and following ones)
+    scripts/config.py unset MBEDTLS_ECP_RESTARTABLE
+
```
The mbedtls version unsets `PSA_WANT_KEY_TYPE_ECC_KEY_PAIR_DERIVE`. Why don't we need this here?

ronald-cron-arm replied:
That's a good point, thanks. I propose to do even more: start from the `test_accel_ecc_all` configuration where there is no ECC builtin crypto (thus unset `PSA_WANT_KEY_TYPE_ECC_KEY_PAIR_DERIVE` but also `MBEDTLS_PK_PARSE_EC_EXTENDED` and `MBEDTLS_PK_PARSE_EC_EXTENDED`). Then just do not accelerate KEY_TYPE_ECC_KEY_PAIR_GENERATE which is expected to enable ecp.c.

davidhorstmann-arm commented on tests/scripts/components-configuration-crypto.sh:
```diff
@@ -91,6 +91,108 @@ component_test_accel_ecc_all_but_ecp_light() {
+    cd $OUT_OF_SOURCE_DIR
+    cmake -DTF_PSA_CRYPTO_TEST_DRIVER=On \
+          -DTF_PSA_CRYPTO_USER_CONFIG_FILE="${user_config_accel_file_path}" ..
+    make
```
Nit: As before, could do `cmake --build .`

davidhorstmann-arm reviewed (changes requested):
A few small comments and questions. Anything marked 'nit' I consider optional.

ronald-cron-arm committed:
user-config-accel*: Remove MBEDTLS_PSA_ACCEL_ECC_SECP_[K|R]1_192

Remove MBEDTLS_PSA_ACCEL_ECC_SECP_[K|R]1_192
as secp192[k|r]1 curves are not supported
anymore.

Signed-off-by: Ronald Cron <ronald.cron@arm.com>

ronald-cron-arm committed:
components-configuration-crypto.sh: Use cmake --build instead of make

Signed-off-by: Ronald Cron <ronald.cron@arm.com>

ronald-cron-arm committed:
test_accel_ecc_some_key_types: Start from no builtin ECC at all

In test_accel_ecc_some_key_types, start from no
builtin ECC at all. Then it is just the fact to
disable the acceleration of
KEY_TYPE_ECC_KEY_PAIR_GENERATE that enables
ecp.c.

Signed-off-by: Ronald Cron <ronald.cron@arm.com>

ronald-cron-arm committed:
psa_crypto_pake.c: Restore MBEDTLS_ERR_MD_FEATURE_UNAVAILABLE translation

Signed-off-by: Ronald Cron <ronald.cron@arm.com>

ronald-cron-arm commented:
@minosgalanakis @davidhorstmann-arm I've pushed four new commits to address David's comments ending up at 5fa5162 and then rebased the 15 commits on top of development to address the conflict (only one comment line) that appeared when #672 was merged.

minosgalanakis reviewed (approved):
LGTM

davidhorstmann-arm reviewed (approved):
LGTM, thanks!
